### PR TITLE
Create SLACKAPI to allow instrument triggering via Slack

### PR DESCRIPTION
**Pull request by mcoughlin**
Slack is a common place to track submission requests, and this PR gives creates a SLACKAPI to allow SP to use the microservice to submit requests to a designated Slack channel.

Blocker: Needs migration

### CODE REVIEW

**Review by profjsb — CHANGES_REQUESTED** [review]
It is important to document (in `docs/`?) how this functionality will be used. It's not clear from the  code how a user/bot gets access to this -- in particular we only enable Slack in the user profile. It's unclear how these requests will enabled (and checked that they are enabled) using API Tokens.   Remember that the Slack interaction is not guaranteed to work -- so should make clear that a request may never go through (e.g. if Slack is down). 

 It's also not clear why specific filters are allowed/disallowed.

I think a detailed description of the usage patterns will help a lot here.

Missing tests.

**Comment by mcoughlin** [comment]
@profjsb These are good questions. Indeed, I would like the SLACKAPI to be able to take in arbitrary filters and so on, so maybe it should be ingesting from whatever the instrument's options are somehow, but maybe it will be tricky to make a generic enough interface...

File: skyportal/facility_apis/slack.py
Comment (mcoughlin) [reply skyportal/facility_apis/slack.py]: Fixed.

File: skyportal/facility_apis/slack.py
Comment (mcoughlin) [reply skyportal/facility_apis/slack.py]: I made it 2 hours.

File: skyportal/facility_apis/slack.py
Comment (mcoughlin) [reply skyportal/facility_apis/slack.py:22]: Added.

File: skyportal/facility_apis/slack.py
Comment (mcoughlin) [reply skyportal/facility_apis/slack.py:71]: I think so... otherwise it's just the app and no one knows who sent it.

File: config.yaml.defaults
Comment (mcoughlin) [reply config.yaml.defaults]: What exactly needs to be enabled here? What is the slack service checking for exactly?

File: data/db_demo.yaml
Comment (mcoughlin) [reply data/db_demo.yaml:305]: Fixed.

File: skyportal/facility_apis/slack.py
Comment (mcoughlin) [reply skyportal/facility_apis/slack.py]: Fixed.

File: doc/slack.md
Comment (profjsb) [inline doc/slack.md:69]: this section should go above `Notes on Reliability and Security`.

File: doc/slack.md
Comment (profjsb) [inline doc/slack.md:71]:
```suggestion
We have enabled the use of Slack for instrument "triggering," enabling a push of a transient follow-up request similar to done for other instruments. To enable this, an instrument needs `SLACKAPI` added as its triggering API, and in the instrument allocation, `_altdata` should look like `{"slack_workspace": "XXX", "slack_channel": "YYY", "slack_token": "ZZZ"}` with the appropriate values. The Slack endpoint will be POSTed to `{SLACK_URL}/{altdata['slack_workspace']}/{altdata['slack_channel']}/{altdata['slack_token']}`
```

**Review by profjsb — CHANGES_REQUESTED** [review]
This will need a migration, as you're adding SLACKAPI to followup_facilities enum. See an example of this in: https://github.com/profjsb/skyportal/blob/90eccb64de9c68621f2ed25f61c9a3e22abca7d9/alembic/versions/f7c1e785d7a4_unique_token_name_by_user.py

**Comment by mcoughlin** [comment]
@profjsb do you understand why the migration checker wasn't triggered on this to give the migra artifact? It worked fine for the observation plan PR which also adds one of these.

**Comment by stefanv** [comment]
No files were touched that would give the workflow a reason to suspect that migrations are involved, so it doesn't run that check.

**Comment by mcoughlin** [comment]
Ah I see. Maybe it could be made to check the init file in facility_apis?

**Comment by profjsb** [comment]
Yes we need to update the GHA action to watch facility APIs

**Comment by mcoughlin** [comment]
@profjsb Yes I am trying it here https://github.com/skyportal/skyportal/pull/2498

**Comment by profjsb** [comment]
I put in a separate PR for watching the files: https://github.com/skyportal/skyportal/pull/2591

**Comment by profjsb** [comment]
Once that gets merged you can rebase and rerun tests and that should trigger the migration manager.

**Comment by mcoughlin** [comment]
yes good. It's not totally trivial to make this migration script work it seems, at least based on the atlas one so far.

**Comment by mcoughlin** [comment]
@profjsb I get migration artifacts that look like the below. Given that the migration script adds SLACKAPI to followup_apis, which (if any) of these lines are relevant to put into the alembic script? All of them?

alter type "public"."followup_apis" rename to "followup_apis__old_version_to_be_dropped";

create type "public"."followup_apis" as enum ('KAITAPI', 'SEDMAPI', 'SEDMV2API', 'IOOAPI', 'IOIAPI', 'SPRATAPI', 'SINISTROAPI', 'SPECTRALAPI', 'FLOYDSAPI', 'MUSCATAPI', 'SLACKAPI', 'ZTFAPI');

alter table "public"."instruments" alter column api_classname type "public"."followup_apis" using api_classname::text::"public"."followup_apis";

drop type "public"."followup_apis__old_version_to_be_dropped";

**Comment by profjsb** [comment]
> @profjsb I get migration artifacts that look like the below. Given that the migration script adds SLACKAPI to followup_apis, which (if any) of these lines are relevant to put into the alembic script? All of them?
> 
> alter type "public"."followup_apis" rename to "followup_apis__old_version_to_be_dropped";
> 
> create type "public"."followup_apis" as enum ('KAITAPI', 'SEDMAPI', 'SEDMV2API', 'IOOAPI', 'IOIAPI', 'SPRATAPI', 'SINISTROAPI', 'SPECTRALAPI', 'FLOYDSAPI', 'MUSCATAPI', 'SLACKAPI', 'ZTFAPI');
> 
> alter table "public"."instruments" alter column api_classname type "public"."followup_apis" using api_classname::text::"public"."followup_apis";
> 
> drop type "public"."followup_apis__old_version_to_be_dropped";

What you're doing in the migration script looks good to me: https://github.com/skyportal/skyportal/blob/9243cc446c878fe6725b0613b68332514f609068/alembic/versions/82f5a89fbe59_slack_trigger_api.py 

migra is detecting a change in the enum and this will (or at least should) satisfy the differences.

**Comment by mcoughlin** [comment]
@profjsb sorry for being dense.. does this mean that file is sufficient or do I need to add those lines into the alembic script?

**Comment by profjsb** [comment]
> @profjsb sorry for being dense.. does this mean that file is sufficient or do I need to add those lines into the alembic script?

No -- those lines above are the alternative route to doing enum updates. Both should work but it's possible migra will only be happy with the verbose way above. Regardless we should not have anything in the `downgrade` path because we dont want to remove enums since that might cause referential integrity issues.

**Review by profjsb — APPROVED** [review]

File: alembic/versions/82f5a89fbe59_slack_trigger_api.py
Comment (stefanv) [inline alembic/versions/82f5a89fbe59_slack_trigger_api.py:27]: My guess here:

The reason this works and not the one above is because you need to do a casting operation of sorts, since the ENUM order changes (note that `'SLACKAPI'` does not get appended, but ends up second from last).

What happens then is that a new replacement enum is created.  This new enum is an entirely different enum, so the database needs to be updated to cast between the old and the new one.  That is what the `using api_classname::text::"public"."followup_apis"` part does—it says change the `api_classname` column to use the new enum and use the values currently in that column to look them up *in the new enum* (the underlying numbers stored on disk will change).

This is probably a somewhat costly operation, but it doesn't seem to have caused problems for us so far.

In the future, the cheapest way to update an enum would be to only *append* values, and then use `ADD VALUE` as Michael did before.

Here's an example, borrowing from the postgres docs, illustrating the above:

```
skyportal=> CREATE TYPE mood AS ENUM ('sad', 'ok', 'happy');
CREATE TYPE
skyportal=> CREATE TABLE person (
    name text,
    current_mood mood
);
CREATE TABLE
skyportal=> INSERT INTO person VALUES ('Moe', 'happy');
INSERT 0 1
skyportal=> SELECT * FROM person WHERE current_mood = 'happy';
 name | current_mood 
------+--------------
 Moe  | happy
(1 row)

skyportal=> alter type mood rename to mood_old;
ALTER TYPE
skyportal=> create type mood as enum ('sad', 'ok', 'not happy', 'happy');
CREATE TYPE
skyportal=> alter table person alter column current_mood type mood;
ERROR:  column "current_mood" cannot be cast automatically to type mood
HINT:  You might need to specify "USING current_mood::mood".
                                                             ^
skyportal=> alter table person alter column current_mood type mood using current_mood::text::mood;
ALTER TABLE
skyportal=> SELECT * FROM person WHERE current_mood = 'happy';
 name | current_mood 
------+--------------
 Moe  | happy
(1 row)

skyportal=> alter type mood add value 'grumpy';
ALTER TYPE
skyportal=> SELECT * FROM person WHERE current_mood = 'happy';
 name | current_mood 
------+--------------
 Moe  | happy
(1 row)
```

File: config.yaml.defaults
Comment (stefanv) [inline config.yaml.defaults:436]: Nitpick, can be ignored:

```suggestion
  followup_requests:
```

File: skyportal/facility_apis/slack.py
Comment (stefanv) [inline skyportal/facility_apis/slack.py:30]:
```suggestion
        request : skyportal.models.FollowupRequest
```

File: skyportal/facility_apis/slack.py
Comment (stefanv) [inline skyportal/facility_apis/slack.py:29]:
```suggestion
```

File: skyportal/facility_apis/slack.py
Comment (stefanv) [inline skyportal/facility_apis/slack.py:35]:
```suggestion
        payload : dict
```

File: skyportal/facility_apis/slack.py
Comment (stefanv) [inline skyportal/facility_apis/slack.py:44]: missing `observation_choices`

File: skyportal/facility_apis/slack.py
Comment (stefanv) [inline skyportal/facility_apis/slack.py:98]:
```suggestion
```

File: skyportal/facility_apis/slack.py
Comment (stefanv) [inline skyportal/facility_apis/slack.py:92]:
```suggestion
```

File: skyportal/facility_apis/slack.py
Comment (stefanv) [inline skyportal/facility_apis/slack.py:103]:
```suggestion
        request : skyportal.models.FollowupRequest
```

File: skyportal/facility_apis/slack.py
Comment (stefanv) [inline skyportal/facility_apis/slack.py:128]: Since these are used in `followup_requests.py` in `post`, the `submit` call has to be made async and awaited, otherwise this will block our handlers.

Your PR is not the one to introduce the issue, so I will file an issue and we can fix it separately.

**Review by stefanv — COMMENTED** [review]
Great work, thanks @mcoughlin.

Approved, pending minor fixes to the docstrings (positions of colons, mainly).  I left a few other nitpicky comments, and you are welcome to use or lose as you choose.

The note about async is a bit more dire, but not your fault so we (I?) can address in another PR.

**Comment by stefanv** [comment]
See also: https://github.com/skyportal/skyportal/issues/2594

File: skyportal/facility_apis/slack.py
Comment (mcoughlin) [reply skyportal/facility_apis/slack.py:43]: It's returned as filters here, but called observation_choices in the code to be consistent with the other APIs. Down the line this will probably evolve if we have spectrographs using this (which have IFU's, and therefore the more generic term makes sense).

File: config.yaml.defaults
Comment (stefanv) [inline config.yaml.defaults:436]: Could you point out to me where this entry is used?

Comment (mcoughlin) [reply config.yaml.defaults]: You are totally right, it works without this.

**Pull request merged**
